pkg/volume/metrics_block.go

```diff
@@ -0,0 +1,81 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package volume
+
+import (
+	"fmt"
+	"io"
+	"os"
+
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var _ MetricsProvider = &metricsBlock{}
+
+// metricsBlock represents a MetricsProvider that detects the size of the
+// BlockMode Volume.
+type metricsBlock struct {
+	// the device node where the volume is attached to.
+	device string
+}
+
+// NewMetricsStatfs creates a new metricsBlock with the device node of the
+// Volume.
+func NewMetricsBlock(device string) MetricsProvider {
+	return &metricsBlock{device}
+}
+
+// See MetricsProvider.GetMetrics
+// GetMetrics detects the size of the BlockMode volume for the device node
+// where the Volume is attached.
+//
+// Note that only the capacity of the device can be detected with standard
+// tools. Storage systems may have more information that they can provide by
+// going through specialized APIs.
+func (mb *metricsBlock) GetMetrics() (*Metrics, error) {
+	metrics := &Metrics{Time: metav1.Now()}
+	if mb.device == "" {
+		return metrics, NewNoPathDefinedError()
+	}
+
+	err := mb.getBlockInfo(metrics)
+	if err != nil {
+		return metrics, err
+	}
+
+	return metrics, nil
+}
+
+// getBlockInfo fetches metrics.Capacity by opening the device and seeking to
+// the end.
+func (mb *metricsBlock) getBlockInfo(metrics *Metrics) error {
+	dev, err := os.Open(mb.device)
+	if err != nil {
+		return fmt.Errorf("unable to open device %q: %w", mb.device, err)
+	}
+	defer dev.Close()
+
+	end, err := dev.Seek(0, io.SeekEnd)
+	if err != nil {
```

Comment:
Have you tested this to work on Windows btw?

Reply:
oops I saw your response to @jsafrane above. Since I don't know anything about Windows, lets check with @jingxu97 or @andyzhangx ?

Reply:
I don't think this would work on Windows, since Windows usually use powershell or dos command for such command. Is it possible to add an empty implementation for Windows?

Reply:
+1 to making two versions of this functions and providing a empty empty implementation for windows. Having said that, it would be weird if Windows did not support seeking a block device to the end via golang APIs. I could be wrong though and hence to be on safe side.

Reply:
found related information on os.seek for Windows
https://github.com/golang/go/issues/36019

seems it does not work if it is a directory handle. But not sure about this case.

Reply:
While researching if `Seek()` works correctly for drives on Windows, I can add a `FIXME` note and skip the detection with a
```go
if runtime.GOOS == "winows" {
    // FIXME: it is uncertain how Windows handles Seek() for block devices
    return
}```

Is that acceptable too?

I am also wondering if it is even possible to provision volumes with `VolumeMode: Block` on Windows?

Reply:
`VolumeMode: Block` is not supported yet on Windows, maybe return error directly for Windows?

Reply:
Thanks @andyzhangx , will return an error and leave a `TODO` comment.

Reply:
This is also done with the latest revision. (Return an error on Windows, not separated functions per OS.)